contracts/SimpleRefundBuilder/RefundBuilderInternal.sol

```diff
@@ -0,0 +1,175 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+import "./RefundBuilderState.sol";
+import "@poolzfinance/poolz-helper-v2/contracts/CalcUtils.sol";
+import "@ironblocks/firewall-consumer/contracts/FirewallConsumer.sol";
+
+/// @title RefundBuilderInternal contract
+/// @notice contain internal logic for Simple Refund Builder
+contract RefundBuilderInternal is RefundBuilderState, FirewallConsumer {
+    using CalcUtils for uint256;
+
+    function _createFirstNFT(
+        ISimpleProvider provider,
+        address token,
+        address owner,
+        uint256 totalAmount,
+        uint256[] memory params,
+        bytes calldata signature
+    ) internal virtual override firewallProtectedSig(0x29454335) returns (uint256 poolId) {
+        // one time token transfer for deacrease number transactions
+        lockDealNFT.mintForProvider(owner, refundProvider);
+        poolId = super._createFirstNFT(provider, token, address(refundProvider), totalAmount, params, signature);
+    }
+
+    function _createFirstNFT(
+        Rebuilder memory data,
+        address from
+    ) internal firewallProtectedSig(0x3da709b8) returns (uint256 poolId){
+        lockDealNFT.mintForProvider(data.userData.userPools[0].user, refundProvider);
+        poolId = lockDealNFT.safeMintAndTransfer(
+            address(refundProvider),
+            data.paramsData.token,
+            from,
+            data.userData.totalAmount,
+            data.paramsData.provider,
+            data.tokenSignature
+        );
+        data.paramsData.provider.registerPool(poolId, data.paramsData.simpleParams);
+    }
+
+    function _createCollateralProvider(
+        address mainCoin,
+        uint256 tokenPoolId,
+        uint256 totalAmount,
+        uint256 mainCoinAmount,
+        uint256 collateralFinishTime,
+        bytes calldata signature
+    ) internal firewallProtectedSig(0x4516d406) returns (uint256 poolId) {
+        poolId = lockDealNFT.safeMintAndTransfer(
+            msg.sender,
+            mainCoin,
+            msg.sender,
+            mainCoinAmount,
+            collateralProvider,
+            signature
+        );
+        uint256[] memory collateralParams = new uint256[](3);
+        collateralParams[0] = totalAmount;
+        collateralParams[1] = mainCoinAmount;
+        collateralParams[2] = collateralFinishTime;
+        collateralProvider.registerPool(poolId, collateralParams);
+        lockDealNFT.cloneVaultId(poolId + 2, tokenPoolId);
+    }
+
+    function _validateParamsData(
+        address[] calldata addressParams,
+        uint256[][] calldata params
+    ) internal view returns (ParamsData memory paramsData) {
+        require(addressParams.length == 3, "invalid addressParams length");
```

Comment:
in this case we can help with : "addressParams must contain exactly 3 addresses"